src/dd/util.rs

```diff
@@ -0,0 +1,180 @@
+            conf.out.buf_size,
+        )
+    };
+    let mut sin;
+    let mut sout;
+    let mut i_f;
+    let mut o_f;
+    let in_: &mut io::Read = if is_inp_std {
+        sin = io::stdin();
+        &mut sin
+    } else {
+        let inp = File::open(&i_name);
+        if inp.is_err() {
+            err(format!("cannot open {}", i_name));
+        }
+        i_f = inp.unwrap();
+        &mut i_f
+    };
+    let out: &mut io::Write = if is_out_std {
+        sout = io::stdout();
+        &mut sout
+    } else {
+        let out = File::create(&o_name);
+        if out.is_err() {
+            err(format!("cannot open {}", o_name));
+        }
+        o_f = out.unwrap();
+        &mut o_f
+    };
+
+    let mut buf = vec![0u8; cmp::min(inp_buf_size, out_buf_size) as usize];
+    let mut total = 0;
+    'read_all_blocks: for _ in 0..count {
+        let mut pos = 0;
+        let mut left = inp_buf_size;
+        let mut session = 0;
+        'read_block: loop {
+            if pos == inp_buf_size {
+                break 'read_block;
+            }
+            let mut is_read_0 = false;
+            match in_.read(&mut buf[pos..pos + left]) {
+                Ok(0) => {
+                    is_read_0 = true;
+                }
+                Ok(num) => {
+                    session += num;
+                    {
+                        let mut conf = cf.lock().unwrap();
+                        total += num;
+                        if num < inp_buf_size {
+                            conf.in_.part += 1;
+                            pos += num;
+                            left = inp_buf_size - pos;
+                        } else if num == inp_buf_size {
+                            conf.in_.full += 1;
+                        } else {
+                            err(format!("overread: {} over {}", num, inp_buf_size));
+                        }
+                    }
+                }
+                Err(_) => {
+                    // TODO: handle
+                }
+            }
+            if session == inp_buf_size || is_read_0 {
+                break 'read_block;
+            }
+        }
+        let num = cmp::min(session, out_buf_size);
+        let mut pos = 0;
+        loop {
+            if pos == num {
+                break;
+            }
+            if let Some(m) = dd_out(&mut buf, pos, num - pos, out) {
+                {
+                    let mut conf = cf.lock().unwrap();
+                    conf.written += m;
+                    if m < conf.out.buf_size {
+                        conf.out.part += 1;
+                    } else {
+                        conf.out.full += 1;
+                    }
+                }
+
+                pos += m;
+            } else {
+                err("error occured".to_string()); // FIXME: make err handling
+            }
+        }
+        if total >= inp_buf_size * count {
+            break;
+        }
+    }
+    display_progress(Arc::clone(&cf));
```

Comment:
I don't think you need to clone here.

Reply:
IMO, It needs to, because display_progress is also used by signal handler, which also gets data in shared pointer.

Reply:
But you're not doing anything with this `Arc` in this function anymore, so can't you just pass ownership to `display_progress()`?

Reply:
display_progress is also called by "main" thread. I made that signature, so that there would be unified interface to both main and called threads.

Reply:
Does the compiler throw an error if you remove the clone?